week-1-ts-ecosystem/w01e04-light-me-up/src/main.ts

```diff
@@ -9,13 +9,59 @@
  * */
 
 class PowerSource {
-  private energySupply = 100
+    private energySupply = 100
 
-  consume(energy) {
-    this.energySupply -= energy
-  }
+    consume(energy: number) {
+        if ((this.energySupply - energy) < 0) {
+            throw new Error('I\'m out of power')
+        }
+
+        this.energySupply -= energy
+    }
+
+    availableEnergy() {
+        return this.energySupply;
+    }
 }
 
 class LightBulb {
-  protected readonly powerConsumption = 20
+    protected readonly powerConsumption = 20
+    powerSource: PowerSource;
+
+    constructor(powerSource: PowerSource) {
+        this.powerSource = powerSource;
+    }
+
+    consume() {
+        this.powerSource.consume(this.powerConsumption);
+    }
+
+    lightOn(sec: number) {
+        for (let s = 0; s < sec; s++) {
+            setTimeout(() => {
+                this.consume();
+                console.log('Light on');
+            }, 1000);
+        }
```

Comment:
@hitch90 - całość zadania jest jak najbardziej ok. poza tym fragmentem.

Zobacz że Twoja żarówka wypali się cała w 1 sekundę
bo dałeś pętle, czyli Twoje 6 timeoutów zostanie wywołanych po 1 sekundzie

żeby to tak zadziałało jak chcesz, musiałoby być `1000 * s`
żeby timeouty fizycznie zadeklarować jako:

`setTimeout(() => ...., 0)`
`setTimeout(() => ...., 1000)`
`setTimeout(() => ...., 2000)`
`setTimeout(() => ...., 3000)`
`setTimeout(() => ...., 4000)`
`setTimeout(() => ...., 5000)`

Tak by to wyglądalo mniej wiecej po "komputacji" jak sobie to wyobrazimy no nie.

No i teraz dochodzimy do wniosku - czy jest sens się w to bawić w ten sposób, skoro jest piękna funkcja:
`setInterval` która sama może dla nas takie coś "odliczyć".

No i dochodzisz do kolejnego wniosku, czy jest sens - ustalać że żarówka ma się świecić dokładnie `6 sec` ?
W sensie, w rzeczywistym świecie, wkręcamy żarówkę - i ona świeci

hmmm ale ile ona świeci....

no poza tym że się może spalić - to świeci tyle ile prądu dostanie.
Wiec jak tutaj wkręcisz żarówke `lightOn` i puścisz `setInterval` co `sec` to Cię już nic nie obchodzi,
bo `PowerSource` rzuci błędem jak się wyczerpie (a to że to się stanie po 6 sekundach wiemy bo obliczyliśmy `powerConsumption` * `ilość sekund zanim się źródło wyczerpie` no nie ?

Rozumiesz co mam na myśli ?